numpy/core/src/multiarray/item_selection.c

```diff
@@ -2128,6 +2129,68 @@ count_nonzero_bytes_384(const npy_uint64 * w)
     return r;
 }
 
+#if NPY_SIMD
+
+/* Count the zero bytes between `*d` and `end`, updating `*d` to point to where to keep counting from. */
+static NPY_INLINE NPY_GCC_OPT_3 npyv_u8
+count_zero_bytes_u8(const npy_uint8 **d, const npy_uint8 *end)
+{
+    const npyv_u8 vone = npyv_setall_u8(1);
+    const npyv_u8 vzero = npyv_zero_u8();
+
+    npy_intp n = 0;
+    npyv_u8 vsum8 = npyv_zero_u8();
+    while (*d < end && n <= 0xFE) {
+        npyv_u8 vt = npyv_cvt_u8_b8(npyv_cmpeq_u8(npyv_load_u8(*d), vzero));
+        vt = npyv_and_u8(vt, vone);
+        vsum8 = npyv_add_u8(vsum8, vt);
+        *d += npyv_nlanes_u8;
+        n += npyv_nlanes_u8;
+    }
+    return vsum8;
+}
+
+static NPY_INLINE npyv_u16
+count_zero_bytes_u16(const npy_uint8 **d, const npy_uint8 *end)
+{
+    npyv_u16 vsum16 = npyv_zero_u16();
+    npy_intp n = 0;
+    while (*d < end && n <= 0xFF00) {
+        npyv_u8 vsum8 = count_zero_bytes_u8(d, end);
+        npyv_u16x2 part = npyv_expand_u16_u8(vsum8);
+        vsum16 = npyv_add_u16(vsum16, npyv_add_u16(part.val[0], part.val[1]));
+        n += 0xFF;
+    }
+    return vsum16;
+}
+
+static NPY_INLINE npyv_u32
+count_zero_bytes_u32(const npy_uint8 **d, const npy_uint8 *end)
+{
+    npyv_u32 vsum32 = npyv_zero_u32();
+    npy_intp n = 0;
+    while (*d < end && n <= 0xFFFF0000) {
+        npyv_u16 vsum16 = count_zero_bytes_u16(d, end);
+        npyv_u32x2 part = npyv_expand_u32_u16(vsum16);
+        vsum32 = npyv_add_u32(vsum32, npyv_add_u32(part.val[0], part.val[1]));
+        n += 0xFFFF;
+    }
+    return vsum32;
+}
+
+static NPY_INLINE npy_intp
+count_nonzero_bytes(const npy_uint8 *d, npy_uintp unrollx)
+{
+    npy_intp zero_count = 0;
+    const npy_uint8 *end = d + unrollx;
+    while (d < end) {
+        npyv_u32 vsum32 = count_zero_bytes_u32(&d, end);
+        zero_count += npyv_sum_u32(vsum32);
```

Comment:
This line currently overflows, as `npyv_sum_u32` results in a vector that contains up to `2**32-1` in each entry, which is then added over multiple entries without any extra carry bits.

Reply:
The maximum value of `unrollx `is `2**31-1`(shape[0] & -16),  if all of the element in d is zero, then `zero_count `will be `2**31-1`,  The results in the final vector contains up to `(2**31-1)/4` in each entry.

Reply:
The vector result of `count_zero_bytes_u32` contains up to `(2**32-1)/4`(eg, 128bits vector) in each entry, because the condition `n <= 0xFFFF0000` is set. the count of zeros will split into each lanes.

Reply:
![image](https://user-images.githubusercontent.com/15245051/102317515-9e730a00-3fb2-11eb-9012-79a9e23dad92.png)
Here is the approximate algorithm process that may explain the overflow problem.